File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.py
def solution(today, terms, privacies):
    today_year, today_month, today_day = map(int, today.split('.'))
    ans = []

    term_dict = {}
    for term in terms:
        t, m = term.split(" ")
        term_dict[t] = int(m)

    for i in range(len(privacies)):
        p_date, p_type = privacies[i].split(" ")
        p_year, p_month, p_day = map(int, p_date.split("."))

        if p_type in term_dict:
            t = term_dict[p_type]
            p_month += t
            p_day -= 1
            
            # 월이 12를 넘어갈 때 처리
            if p_month > 12:
                p_year += (p_month - 1) // 12
                p_month = (p_month - 1) % 12 + 1

        if (p_year < today_year) or \
           (p_year == today_year and p_month < today_month) or \
           (p_year == today_year and p_month == today_month and p_day < today_day):
            ans.append(i + 1)

    return ans

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.js
function solution(today, terms, privacies) {
    var answer = [];
    let [tYear, tMonth, tDay] = today.split(".").map(Number);
    for (let i = 0; i < privacies.length; i++){
        let [privacieseDate, privaciesType] = privacies[i].split(" ");
        let [pYear, pMonth, pDay] = privacieseDate.split(".").map(Number);
        for (let j = 0; j < terms.length; j++){
            let [termsType, termsMonth] = terms[j].split(" ");
            termsMonth = Number(termsMonth);
            if (termsType === privaciesType){
                pMonth += termsMonth;
                pYear += Math.floor((pMonth - 1) / 12);
                pMonth = ((pMonth - 1) % 12) + 1;
            }
        }
        pDay -= 1;
        if (pDay === 0) {
            pDay = 28;
            pMonth -= 1;
            if (pMonth === 0) {
                pMonth = 12;
                pYear -= 1;
            }
        }
        if (tYear > pYear){
            answer.push(i+1);
            continue;
        }
        if (tYear === pYear && tMonth > pMonth){
            answer.push(i+1);
            continue;
        }
        if (tYear === pYear && tMonth === pMonth && tDay > pDay){
            answer.push(i+1);
            continue;
        }
    }
    return answer;
}